File: src/aircraft/game/supply/Supply.java
package aircraft.game.supply;

import java.awt.geom.Point2D;
import java.awt.Graphics;
import java.awt.image.BufferedImage;

import aircraft.game.Canvas;
import aircraft.game.ImageLoader;
import aircraft.game.plane.Plane;

public class Supply {
  protected Canvas canvas;
  public BufferedImage image;

  // The current position.
  public Point2D.Double location = new Point2D.Double();
  protected double speed = 1.5; // The speed of the flying object.
  private int recovery = 20;

  public Supply(Canvas canvas, double x, double y) {
    // Load the plane from the image directory.
    this.canvas = canvas;
    image = ImageLoader.readImg("supply.gif");
    if (image.getWidth()  >= canvas.getWidth() ||
        image.getHeight() >= canvas.getHeight())
      throw new RuntimeException("The size of image is invalid");
    
    location.x = x; location.y = y;
  }

  // The display method: as the name suggests, this method draw the image
  // onto the graphics user interface, so the argument is necessary.
  public void display(Graphics graphics) {
    // Default implementation: show image on the gui.
    graphics.drawImage(image, (int)location.x, (int)location.y, null);
  }

  // Check whether the supply is displayed inside the window.
  public void boundaryCheck() {
    double xmin = location.x; // Left boundary.
    double xmax = location.x + image.getWidth(); // Right boundary.
    double ymin = location.y; // Bottom boundary.
    double ymax = location.y + image.getHeight(); // Top boundary.

    if (xmin < 0 || xmax > canvas.getWidth() ||
        ymin < 0 || ymax > canvas.getHeight())
      canvas.trash.add(this);
  }

  // Each object should have a specific move policy.
  public void move() {
    location.y += speed;
  }

  public void effect(Plane plane) {
    plane.health += this.recovery;
  }
}

File: python/supply.py
from point2d import Point2D
import imgloader
import setting

class Supply:
    def __init__(self, canvas, x, y):
        """Initialize the supply class.

        Keyword arguments:
        canvas -- the canvas for gui.
        x -- the x coordinate of location.
        y -- the y coordinate of location.
        """
        # Read image and check its size.
        self.canvas = canvas
        self.image = imgloader.load(setting.IMAGES["Supply"])
        w, h = self.image.get_rect().size
        if w >= self.canvas.width or h >= self.canvas.height:
            raise ValueError("The size of image is invalid")

        self.location = Point2D(x, y)
        self.speed = setting.SPEED["Supply"] # The speed of the flying object.
        self._recovery = setting.RECOVERY["Supply"]

    def display(self, graphics):
        """Display image on the screen.
        """
        # Show image on the screen with respect to the specified location
        # and the image sources. Note that this method is not abstract.
        graphics.blit(self.image, self.location.cartesian())

    def boundary_check(self):
        """Check whether the supply moves out of window.

        One supply will be directly moved to trash and deleted in the next frame
        once it moves out of the current game window.
        """
        w, h = self.image.get_rect().size
        if self.location.x < 0 or self.location.x + w > self.canvas.width or \
           self.location.y < 0 or self.location.y + h > self.canvas.height:
            self.canvas.objects["trash"].add(self)

    def move(self):
        """Move the supply to the next location at the next frame, according to
        the current speed and direction.
        """
        self.location.y += self.speed

    def effect(self, plane):
        plane.health += self._recovery

    def camp(self):
        """Return the camp of this object.
        """
        return setting.NEUTRAL
